### PR TITLE
Add TCAHighlightConverter

### PR DESCRIPTION
**Pull request by ichttt**
As discussed in  #13, this is an alternative approach that teaches log4j about the TCA terminal state for ANSI coloring.
Due to `isAnsiSupported` being called early this way, I needed to add a call to `initializeTerminal` in this function. I optimised the function so it doesn't need to synchronize if the initilized flag is already set.
This isn't a perfect approach as well, but it works and I guess it is less of a workaround then #13 
Feedback welcome.

### CODE REVIEW

**Comment by stephan-gh** [comment]
Sorry for never replying to this. I kind of like this one actually, it's fairly simple yet allows to use Log4j's highlight converter pretty much as-is. Do you (or someone else) still need this functionality @ichttt?

**Comment by ichttt** [comment]
Yes this would still be super nice. I implemented this in forge as well (https://github.com/MinecraftForge/MinecraftForge/blob/1.16.x/src/fmllauncher/java/net/minecraftforge/fml/loading/log4j/ForgeHighlight.java) but it is obviously more hacky there as I need to call a private function. I would be nice if we could get rid of this in forge and just bump TCA

File: src/main/java/net/minecrell/terminalconsole/TCAHighlightConverter.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
+package net.minecrell.terminalconsole;
```
Comment (stephan-gh) [inline src/main/java/net/minecrell/terminalconsole/TCAHighlightConverter.java:25]: Can you run `./gradlew licenseFormat`?

File: src/main/java/net/minecrell/terminalconsole/TCAHighlightConverter.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import org.apache.logging.log4j.util.PerformanceSensitive;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * A wrapped version of {@link HighlightConverter} that uses
+ * {@link TerminalConsoleAppender} to detect if Ansi escape codes can be used
```
Comment (stephan-gh) [inline src/main/java/net/minecrell/terminalconsole/TCAHighlightConverter.java]:
```suggestion
 * {@link TerminalConsoleAppender} to detect if ANSI escape codes can be used
```

File: src/main/java/net/minecrell/terminalconsole/TerminalConsoleAppender.java [hunk 3]
```diff
@@ -205,44 +206,48 @@ protected TerminalConsoleAppender(String name, Filter filter, Layout<? extends S
         initializeTerminal();
     }
 
-    private synchronized static void initializeTerminal() {
```
Comment (stephan-gh) [inline src/main/java/net/minecrell/terminalconsole/TerminalConsoleAppender.java]: I think I would prefer if you keep this method as-is and rather change the calls of `initializeTerminal();` to

```java
if (!initialized)
    initializeTerminal();
```

AFAICT there are just two calls of `initializeTerminal();` so the code overhead is negligible. I personally don't like the explicit `synchronized (TerminalConsoleAppender.class) {` very much.

File: src/main/java/net/minecrell/terminalconsole/TCAHighlightConverter.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+ * It can be only used together with
+ * {@link TerminalConsoleAppender}.</p>
+ *
+ * <p>{@link TerminalConsoleAppender#ANSI_OVERRIDE_PROPERTY} may be used
+ * to force the use of ANSI colors even in unsupported environments.</p>
+ *
+ * <p><b>Example usage:</b> {@code %TCAhighlight{%level: %message}}</p>
+ */
+@Plugin(name = "TCAhighlight", category = PatternConverter.CATEGORY)
+@ConverterKeys({ "TCAhighlight" })
```
Comment (stephan-gh) [inline src/main/java/net/minecrell/terminalconsole/TCAHighlightConverter.java]:
```suggestion
@ConverterKeys({ "highlightTCA" })
```

would be a bit easier to read if you ask me. What do you think?

**Review by stephan-gh — COMMENTED** [review]
Great! Just a few minor comments :)

File: src/main/java/net/minecrell/terminalconsole/TerminalConsoleAppender.java (hunk 3 above)
Comment (ichttt) [reply src/main/java/net/minecrell/terminalconsole/TerminalConsoleAppender.java]: It is used more in HighlightErrorConverter, but sure, the overhead should still be very small

Comment (stephan-gh) [reply src/main/java/net/minecrell/terminalconsole/TerminalConsoleAppender.java]: `HighlightErrorConverter` doesn't call `initializeTerminal()` directly though. What I mean is that you change `isAnsiSupported()` to

```java
public static boolean isAnsiSupported() {
        if (!initialized)
             initializeTerminal();
        return ANSI_OVERRIDE != null ? ANSI_OVERRIDE : terminal != null;
    }
```

and same for the constructor

**Review by stephan-gh — APPROVED** [review]

**Comment by stephan-gh** [comment]
@ichttt Thanks a lot for making this PR. I made the change I suggested above just to avoid potential performance regressions (see https://github.com/Minecrell/TerminalConsoleAppender/commit/554276543bf1ecbc86e9d61f26a67e1c6058a295).

Please make sure that the `%highlightTCA` works the way you need it (I don't need it myself in any project so I won't test it). If you want, you can also help testing the JLine 3.12.0 upgrade for Forge as described in https://github.com/Minecrell/TerminalConsoleAppender/issues/19. :)

**Comment by stephan-gh** [comment]
Actually, I just tested it briefly for https://github.com/Minecrell/TerminalConsoleAppender/commit/e4713d0c784837c0019fdbd22baecae1401fc167. Works great, finally Log4j understands when to disable ANSI. Thanks again :)

**Pull request merged**